src/jrd/replication/Publisher.cpp

```diff
@@ -0,0 +1,635 @@
+#include "../jrd/Database.h"
+#include "../jrd/blb_proto.h"
+#include "../jrd/cch_proto.h"
+#include "../jrd/evl_proto.h"
+#include "../jrd/met_proto.h"
+#include "../jrd/mov_proto.h"
+#include "../common/isc_proto.h"
+
+#include "Publisher.h"
+#include "Replicator.h"
+
+using namespace Firebird;
+using namespace Jrd;
+using namespace Replication;
+
+namespace
+{
+	// Generator RDB$BACKUP_HISTORY, although defined as system,
+	// should be replicated similar to user-defined ones
+	const int BACKUP_HISTORY_GENERATOR = 9;
+
+	const char* LOG_ERROR_MSG = "Replication is stopped due to critical error(s)";
+
+ 	void handleError(thread_db* tdbb, jrd_tra* transaction = NULL)
+	{
+		const auto dbb = tdbb->getDatabase();
+		fb_assert(dbb);
+		const auto attachment = tdbb->getAttachment();
+		fb_assert(attachment);
+
+		if (transaction && transaction->tra_replicator)
+		{
+			transaction->tra_replicator->dispose();
+			transaction->tra_replicator = NULL;
+		}
+
+		if (attachment->att_replicator)
+		{
+			const auto status = attachment->att_replicator->getStatus();
+
+			if (status->getState() & IStatus::STATE_ERRORS)
+			{
+				Firebird::string msg;
+				msg.printf("Database: %s\n\t%s", dbb->dbb_filename.c_str(), LOG_ERROR_MSG);
+				iscLogStatus(msg.c_str(), status);
+			}
+
+			attachment->att_replicator->dispose();
+			attachment->att_replicator = NULL;
+		}
+	}
+
+	Record* upgradeRecord(thread_db* tdbb, jrd_rel* relation, Record* record)
+	{
+		const auto format = MET_current(tdbb, relation);
+
+		if (record->getFormat()->fmt_version == format->fmt_version)
+			return record;
+
+		auto& pool = *tdbb->getDefaultPool();
+		const auto newRecord = FB_NEW_POOL(pool) Record(pool, format);
+
+		dsc orgDesc, newDesc;
+
+		for (auto i = 0; i < newRecord->getFormat()->fmt_count; i++)
+		{
+			newRecord->clearNull(i);
+
+			if (EVL_field(relation, newRecord, i, &newDesc))
+			{
+				if (EVL_field(relation, record, i, &orgDesc))
+					MOV_move(tdbb, &orgDesc, &newDesc);
+				else
+					newRecord->setNull(i);
+			}
+		}
+
+		return newRecord;
+	}
+
+	IReplicatedTransaction* ensureTransaction(thread_db* tdbb, jrd_tra* transaction)
+	{
+		// Create a transaction object, unless it already exists
+
+		if (!transaction->tra_replicator)
+		{
+			const auto attachment = tdbb->getAttachment();
+			fb_assert(attachment);
+
+			const auto replicator = attachment->att_replicator;
+
+			if (!replicator)
+				return NULL;
+
+			transaction->tra_replicator = replicator->startTransaction(transaction->tra_number);
+
+			if (!transaction->tra_replicator)
+				handleError(tdbb);
+		}
+
+		return transaction->tra_replicator;
+	}
+
+	bool ensureSavepoints(thread_db* tdbb, jrd_tra* transaction)
+	{
+		const auto dbb = tdbb->getDatabase();
+		fb_assert(dbb);
+
+		const auto replicator = transaction->tra_replicator;
+
+		// Replicate the entire stack of active savepoints (excluding priorly replicated),
+		// starting with the oldest ones
+
+		HalfStaticArray<Savepoint*, 16> stack;
```

Comment:
This stack has no point: because of "relative" savepoint handling you need only count of them but not order. On success result doesn't matter and in case of any error complete replicator is destroyed and "replicated" flag in savepoint is irrelevant.

Reply:
Agreed about the count only (the very-first implementation was passing savepoint IDs, the stack still remains). The "replicated" flag is relevant to avoid starting priorly started (upper level) savepoints. Imagine the first update executing inside savepoint frame {1, 2, 4, 10}, then savepoint 10 is released and the next update is executed inside savepoint frame {1, 2, 4, 12}. Savepoints {1, 2, 4} were already started and should not be replicated again.

Reply:
I said that it is irrelevant in the case if error happen.
No error retirned from startSavepoint() - whole savepoint stack is marked as "replicated" and no problem here.
On error all savepoints are rolled back and dropped.